src/python/pants/core/goals/package.py

```diff
@@ -178,5 +185,35 @@ async def package_asset(workspace: Workspace, dist_dir: DistDir) -> Package:
     return Package(exit_code=0)
 
 
+@dataclass(frozen=True)
+class TraverseIfNotPackageTarget(ShouldTraverseDepsPredicate):
+    package_field_set_types: FrozenOrderedSet[PackageFieldSet]
+    roots: FrozenOrderedSet[Address]
+    always_traverse_roots: bool = True  # traverse roots even if they are package targets
+
+    def __init__(
+        self,
+        *,
+        union_membership: UnionMembership,
+        roots: Iterable[Address],
+        always_traverse_roots: bool = True,
+    ) -> None:
+        object.__setattr__(self, "package_field_set_types", union_membership.get(PackageFieldSet))
+        object.__setattr__(self, "roots", FrozenOrderedSet(roots))
+        object.__setattr__(self, "always_traverse_roots", always_traverse_roots)
+        self.__post_init__()
```

Comment:
this is a huuge foot gun.

The sensible thing imo would be to change the base class to use `__init__` rather than `__post_init__` so this could become:
```suggestion
        super().__init__()
```

Calling super makes sense, calling post init is not something I would think of doing.

....

Just realized, can still call super init here I think, as the parent init will call post init for us.

Reply:
Oh. I like calling super a lot better!